prompt_toolkit/key_binding/bindings/emacs.py

```diff
@@ -364,3 +370,150 @@ def _(event: E) -> None:
             count=event.arg)
 
     return ConditionalKeyBindings(key_bindings, emacs_mode)
+
+
+def load_emacs_shift_selection_bindings() -> KeyBindingsBase:
+    """
+    Bindings to select text with shift + cursor movements
+    """
+
+    key_bindings = KeyBindings()
+    handle = key_bindings.add
+
+    def unshift_move(event: E) -> None:
+        """
+        Used for the shift selection mode. When called with
+        a shift + movement key press event, moves the cursor
+        as if shift is not pressed.
+        """
+        key = event.key_sequence[0].key
+
+        if key == Keys.ShiftUp:
+            event.current_buffer.auto_up(count=event.arg)
+            return
+        if key == Keys.ShiftDown:
+            event.current_buffer.auto_down(count=event.arg)
+            return
+
+        # the other keys are handled through their readline command
+        key_to_command = {
+            Keys.ShiftLeft:         'backward-char',
+            Keys.ShiftRight:        'forward-char',
+            Keys.ShiftHome:         'beginning-of-line',
+            Keys.ShiftEnd:          'end-of-line',
+            Keys.ShiftControlLeft:  'backward-word',
+            Keys.ShiftControlRight: 'forward-word',
+            Keys.ShiftControlHome:  'beginning-of-buffer',
+            Keys.ShiftControlEnd:   'end-of-buffer'
+        }
+
+        try:
+            # Both the dict lookup and `get_by_name` can raise KeyError.
+            handler = get_by_name(key_to_command[key])
+        except KeyError:
+            pass
+        else:  # (`else` is not really needed here.)
```

Comment:
Actually, `else` is needed here, because we don't want `handler` to execute if there was a `KeyError`.
(I can remove the comment in a separate commit.)